src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java

```diff
@@ -93,10 +95,24 @@
     private Channel controlMessageChannel;
 
     // note: this really doesn't need to be a LBQ, just something that's thread safe
-    private Collection<ScheduledFuture<?>> channelKeepAlives = new LinkedBlockingQueue<>();
+    private final Collection<ScheduledFuture<?>> channelKeepAlives = new LinkedBlockingQueue<>();
 
     private final ThreadPoolExecutor fileTransferExecutor;
 
+    /**
+     * A {@link ThreadLocal} used by the threads in {@link #fileTransferExecutor} to stash references to constructed
+     * and connected {@link Channel}s.
+     */
+    private static final FastThreadLocal<Channel> threadLocalChannel = new FastThreadLocal<>();
+
+    /**
+     * A netty channel attribute used to indicate if a channel is currently transferring a file. This is primarily used
```

Comment:
But now the issue is that's it's racey :-) Keep alive needs to hold a lock while checking the value and sending the message. Streaming needs hold the lock while setting the value.